### ISSUE TITLE
[Docos] Add a new FAQ entry instructing everybody to ignore `typer` :face_holding_back_tears: 

### ISSUE THREAD

**Issue by leycec** [beartype/beartype#436]
Let's officially support `typer`, an increasingly popular annotations-centric CLI framework. Sadly, `typer` violates PEP standards in a few obvious ways. But... *who really cares!?* `typer` is nifty. Nifty packages deserve our support. Let's do this when we get a spare free moment.

@taranlu-houzz over at discussion #434 has [all the spicy deets](https://github.com/beartype/beartype/discussions/434#discussioncomment-10707838):

> As far as the `typer` issue goes, it mainly has to do with how one turns normal functions into `typer` commands. Since `typer` also does validation and some other complex work based on your function params, it is often necessary to use the `typer.Option` or `typer.Argument` objects, which triggers `beartype` if you try to use the function (that had been decorated by `typer`) with that actual expected values that would be passed to the function.

```python
@APP.command()
def sync(
    targets: list[int] = typer.Argument(
        ...,
        metavar="PID [PID [...]]",
        help="Target PIDs to sync tile data for.",
    ),
    submit_change: bool = typer.Option(
        False,
        rich_help_panel="Perforce Options",
        help="Whether or not to submit the changelist with the downloaded tile data (if any).",
    ),
) -> None:
    ...
```

> So, for example, the `submit_change` param expects a `typer.Option` object as far as `beartype` is concerned, even though it would be reasonable to call this function by passing a `bool` directly.

## What Can Be Done Until @leycec Does Something, Though?

Thankfully, you don't have to wait for @leycec to finally do something. Instead, just tell @beartype that it should ignore all `@APP.command()`-decorated callables like so:

```python
from typing import no_type_check

@APP.command()
@no_type_check
def sync(
    targets: list[int] = typer.Argument(
        ...,
        metavar="PID [PID [...]]",
        help="Target PIDs to sync tile data for.",
    ),
    submit_change: bool = typer.Option(
        False,
        rich_help_panel="Perforce Options",
        help="Whether or not to submit the changelist with the downloaded tile data (if any).",
    ),
) -> None:
    ...
```

Like all runtime type-checkers (e.g., `typeguard`) and static type-checkers (e.g., `mypy`, `pyright`), @beartype silently ignores *all* callables and classes decorated by the standard `@typing.no_type_check` decorator. Since `typer` kinda violates PEP standards, <sup>*...sad!*</sup> it's helpful to notify type-checkers that they should ignore `typer`-based callables.

@beartype + `typer`: *soon to be a match made in CLI heaven.* :angel:

**Comment by taranlu-houzz** [comment]
@leycec I haven't tested it yet, but I realized that I was using the "old" way of doing things when adding the `typer` arguments and options. There is a newer style that uses `Annotated`, which seems like it might actually resolve this issue. I will test that out in the near future.

**Comment by leycec** [comment]
**Superb!** Indeed, `typing.Annotated[...]` *should* solve almost everything. You'll probably still need to strip ANSI escape sequences like `[0m` from logged type-checking violation messages by disabling @beartype colourization: e.g.,

```python
# Please do either this...
@beartype(conf=BeartypeConf(is_color=False))

# ...or this, depending on which you're running with now!
beartype_this_package(conf=BeartypeConf(is_color=False))
```

Thanks so much for sticking with @beartype through all this. I swear this will work someday. :smiling_face_with_tear:

**Comment by taranlu-houzz** [comment]
@leycec I have confirmed that `Annotated` does seem to resolve my issue here, which is great! I do have one more quick question: I would like to take advantage of the nicely formatted rich exceptions in the same way that `typer` does when using my package as a library. Is there a way to hook this up to the logger that `beartype` uses: https://rich.readthedocs.io/en/latest/logging.html#handle-exceptions

**Comment by leycec** [comment]
**Wondrousness!** Apologies for my *extreme* delay, too. I've been putting out raging fires all over the issue tracker for the last week. My excuses may be weak, but they're all I have. :smile: 

So. Let's see here...

> Is there a way to hook this up to the logger that `beartype` uses...

Interestingly, @beartype currently logs *nothing*. @beartype doesn't print anything, either! All @beartype does is:

* **Raise fatal exceptions.** *All* exceptions raised by @beartype are standard pure-Python exceptions that subclass the standard `Exception` superclass. In theory, the `RichHandler`-based logging configuration you've so kindly linked to should already transparently handle @beartype exceptions in the same exact way as everybody else's exceptions. *In theory.* :wink: 
* **Emit non-fatal warnings.** This is probably the part that's tripping you up, right? By default, I believe Python prints unhandled non-fatal warnings to **standard error.** If you'd like to intervene in that process and log *all* warnings (including @beartype warnings) with `rich` instead, [a trivial one-liner should do you](https://docs.python.org/3/library/logging.html#integration-with-the-warnings-module):

```python
from logging import captureWarnings
captureWarnings(True)
```

That's... it. I think? Admittedly, Python logging gets *really* ugly really fast. If the above still fails to suffice for you, see [this `discuss.python.org` thread for a demonstration of Python logging getting *really* ugly really fast](https://discuss.python.org/t/some-easy-and-pythonic-way-to-bind-warnings-to-loggers/14009).

Alternately, consider submitting an issue to the upstream `rich` issue tracker. Although the official `rich` documentation presents trivial code for handling uncaught exceptions (which *should* include uncaught @beartype exceptions), their documentation mostly ignores warnings. But warnings are *almost* as mission critical as exceptions, right?

Let's stroke our stubbly chins thoughtfully. :thinking:

**Comment by leycec** [comment]
**Urgh!** I briefly looked into adding official `typer` support to @beartype tonight. Sadly, it's infeasible from @beartype's perspective. I know, right? I'm sobbing over here. :sob: 

This is a core `typer` issue. @beartype can't do much about this without forcibly monkey-patching the `typer` codebase – *which @beartype is not going to do.* The `@typer.Typer.command()` decorator just trivially returns the decorated function unmodified. Since `@typer.Typer.command()` doesn't even modify the function it decorates, @beartype has no means of detecting that the function it's decorating was previously decorated by `@typer.Typer.command()`.

## What `typer` Is Doing: It Kinda Sucks

Specifically, this is what `@typer.Typer.command()` is doing:

```python
    def command(
        self,
        name: Optional[str] = None,
        *,
        cls: Optional[Type[TyperCommand]] = None,
        context_settings: Optional[Dict[Any, Any]] = None,
        help: Optional[str] = None,
        epilog: Optional[str] = None,
        short_help: Optional[str] = None,
        options_metavar: str = "[OPTIONS]",
        add_help_option: bool = True,
        no_args_is_help: bool = False,
        hidden: bool = False,
        deprecated: bool = False,
        # Rich settings
        rich_help_panel: Union[str, None] = Default(None),
    ) -> Callable[[CommandFunctionType], CommandFunctionType]:
        if cls is None:
            cls = TyperCommand

        def decorator(f: CommandFunctionType) -> CommandFunctionType:
            self.registered_commands.append(
                CommandInfo(
                    name=name,
                    cls=cls,
                    context_settings=context_settings,
                    callback=f,
                    help=help,
                    epilog=epilog,
                    short_help=short_help,
                    options_metavar=options_metavar,
                    add_help_option=add_help_option,
                    no_args_is_help=no_args_is_help,
                    hidden=hidden,
                    deprecated=deprecated,
                    # Rich settings
                    rich_help_panel=rich_help_panel,
                )
            )
            return f  # <--- *THIS IS THE PROBLEM, YO*

        return decorator
```

See the `return f` statement I've helpfully commented? Indeed, that is the problem, yo.

## What `typer` Should Be Doing: It No Longer Sucks

To play nicely with other decorators like @beartype, `typer` needs to be adding some sort of `typer`-specific metadata to the returned function `f`. In particular, this is what @beartype *really* wants:

```python
        def decorator(f: CommandFunctionType) -> CommandFunctionType:
            command_info = CommandInfo(
                name=name,
                cls=cls,
                context_settings=context_settings,
                callback=f,
                help=help,
                epilog=epilog,
                short_help=short_help,
                options_metavar=options_metavar,
                add_help_option=add_help_option,
                no_args_is_help=no_args_is_help,
                hidden=hidden,
                deprecated=deprecated,
                # Rich settings
                rich_help_panel=rich_help_panel,
            )

            self.registered_commands.append(command_info)
            f.__typer_command_info__ = command_info  # <--- *THIS IS THE SOLUTION, YO*
            return f
```

@beartype wants `@typer.Typer.command()` to set a `typer`-specific `__typer_command_info__` dunder attribute on the returned callable containing the relevant `CommandInfo` object.

The whole thing's a bit oddball, honestly. The way `typer` is decorating callables is *definitely* weird, non-standard, and hostile to other decorators. Even the `__typer_command_info__` approach isn't that great.

The best possible approach would be for `typer` to create and return a *new* decorated closure wrapped by the standard `@functools.wraps(f)` decorator, which then propagates metadata (especially annotations) from the original callable onto the new closure. But... I just can't see `typer` doing that. `typer` is so far off the map here that even the `__typer_command_info__` approach would be a welcome miracle.

@beartype has fallen down and can't get up. Since `typer` doesn't play nicely with other decorators, @beartype's paws are tied here. Let's reluctantly close this issue and pray for better commits from `typer`.

If `typer` ever resolves this for @beartype (and other decorators), just let me know. I'll happily reopen and resolve this properly. Until then, I hang my neckbeard in shame. :face_exhaling:

**Comment by leycec** [comment]
Actually, @beartype *can* do something. @beartype can loudly advise everybody to stop type-checking PEP-noncompliant `typer` commands with a new FAQ entry. @beartype users are strong. They can handle the truth! :neckbeard:

**Comment by leycec** [comment]
**...lol.** I am frequently wrong. We can admit that – some of us all too readily.

Clearly, requiring users to manually hack around (arguable) deficiencies in @beartype is awful API design. Typer is popular. @beartype needs to transparently support Typer. It's that simple.

## It's Not That Simple

To support Typer, @beartype now needs to support a new category of third-party **runtime-hostile decorators.** These are decorators so hostile to runtime Python environments that you can't even detect them. They're the Lyme Disease of the decorator space. You can't see them. You can't fight them. They're the invisible, silent killer of inter-package compatibility.

Can @beartype cope? Yes – but just barely. `beartype_this_package()` elevates @beartype into a hybrid static-runtime type-checker. We can do whatever we're willing to waste enough time doing! By leveraging the horrifying power of static analysis and abstract syntax tree (AST) transformations at runtime, `beartype_this_package()` can at static analysis (i.e., module import) time:

1. Detect attempts to import the runtime-hostile `@typer.Typer.command` decorator.
2. Detect attempts to decorate callables with that decorator.
3. Prefix all such decorations with the standard `@typing.no_type_check` decorator.

`beartype_this_package()` will thus dynamically transform this minimal-length example that hates the `@beartype` decorator:

```python
import typer

app = typer.Typer()

@app.command()
def main(name: str):
    print(f"Hello {name}")

if __name__ == "__main__":
    app()
```

...into this equivalent example that actually works:


```python
import typer
from typing import no_type_check  # <-- first trivial change: it is good

app = typer.Typer()

@no_type_check  # <-- next trivial change: it is even gooder
@app.command()
def main(name: str):
    print(f"Hello {name}")

if __name__ == "__main__":
    app()
```

It sounds easy – but so did regrowing my hair. It's not easy. This is my uneasy plan:

1. On the first importation of literally *anything* from the `typer` package, `beartype_this_package()` will implicitly inject the `from typing import no_type_check` import immediately after that `typer` import.
1. On each instantiation of the `typer.Typer` class, `beartype_this_package()` will record the variable to which that instantiation is assigned.
1. Lastly, on each decoration calling the `command()` method of such a variable, `beartype_this_package()` will inject the `@no_type_check` decorator.

## Honestly... This Sucks

I know, right? Am I *really* gonna do all that just to ignore PEP-noncompliant Typer stuff? No idea. It's theoretically feasible, at least. It doesn't require monkey-patching Typer and *should* work in the common case. The trouble with AST transformations, of course, is that they tend to break down in uncommon edge cases. But... does anybody care about those? They'd better not!

Our patience has worn thin with you, Typer. Why are you like this... 😭

**Comment by leycec** [comment]
**UPDATE:** Please see the megathread at #542 for further discussion and this temporary hot-fix for everybody:

```python
from beartype import BeartypeConf, BeartypeDecorationPosition
from beartype.claw import beartype_this_package

beartype_this_package(conf=BeartypeConf(
    claw_decoration_position_funcs=BeartypeDecorationPosition.FIRST))
```

💪 🐻

**Comment by leycec** [comment]
**tl;dr:** Typer type hints that look like `typing.Annotated[str, typer.Option(help="Last name of person to greet.")]` instead of `typer.Option(help="Last name of person to greet.")` is the answer in 2025. Continue reading for a boring manologue that will make your hair fall out.

## Boring Manologue: It May Cost You Your Hair

**...lol.** I *really* wanted to add official support for the `@typer.Typer.command` decorator to @beartype 0.22.0. Unfortunately, I'm all outta time and bubblegum. If I don't go back in time and release @beartype 0.22.0 yesterday, other packages start breaking. Python 3.14 is almost here. We're up against the release wall. 😬 

Fortunately, it's unclear whether this issue is even relevant in 2025. Like so many things (e.g., my luxurious hair, my toned physique), Typer has moved on. Why? How? Because [Typer now officially supports PEP 593-compliant `typing.Annotated[...]` type hints](https://typer.tiangolo.com/tutorial/options/help/). That is *absolutely* what every Typer user in the world should now be using. They look like this:

```python
import typer
from typing import Annotated


def main(
    name: str,
    lastname: Annotated[str, typer.Option(help="Last name of person to greet.")] = "",
    formal: Annotated[bool, typer.Option(help="Say hi formally.")] = False,
):
    """
    Say hi to NAME, optionally with a --lastname.

    If --formal is used, say hi very formally.
    """
    if formal:
        print(f"Good day Ms. {name} {lastname}.")
    else:
        print(f"Hello {name} {lastname}")


if __name__ == "__main__":
    typer.run(main)
```

@beartype was correct to complain about Typer's prior use of PEP-noncompliant type hints – and still is. In theory, this means that @beartype and Typer are now BFFLs again. We used to hate each other. Now we quietly tolerate each other. Friendship in the digital age is like that. 😂 

Since Typer now plays ball with PEP standards, @beartype now plays ball with Typer. That means this issue no longer has a reason to live, though. If anyone objects, please re-open this and I'll happily take a deeper, harder, longer look at Typer. Until then, thanks so much for playing along with @beartype.

All your CLI belong to Typer. 🫥
